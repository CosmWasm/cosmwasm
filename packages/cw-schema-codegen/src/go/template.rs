use askama::Template;
use std::borrow::Cow;

#[derive(Clone)]
pub struct EnumVariantTemplate<'a> {
    pub name: Cow<'a, str>,
    pub rename: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
}

#[derive(Template)]
#[template(escape = "none", path = "go/enum.tpl.go")]
pub struct EnumTemplate<'a> {
    pub add_package: bool,
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub variants: Cow<'a, [EnumVariantTemplate<'a>]>,
    pub has_unit_variants: bool,
}

#[derive(Clone)]
pub struct FieldTemplate<'a> {
    pub name: Cow<'a, str>,
    pub rename: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: Cow<'a, str>,
}

#[derive(Clone)]
pub enum TypeTemplate<'a> {
    Unit,
    Tuple(Cow<'a, [Cow<'a, str>]>),
    Named {
        fields: Cow<'a, [FieldTemplate<'a>]>,
    },
}

impl TypeTemplate<'_> {
    pub fn is_unit(&self) -> bool {
        matches!(self, TypeTemplate::Unit)
    }
}

#[derive(Template)]
#[template(escape = "none", path = "go/struct.tpl.go")]
pub struct StructTemplate<'a> {
    pub add_package: bool,
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
}
