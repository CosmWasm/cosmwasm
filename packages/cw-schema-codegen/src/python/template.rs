use askama::Template;
use std::borrow::Cow;

#[derive(Clone)]
pub struct EnumVariantTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
}

#[derive(Template)]
#[template(escape = "none", path = "python/enum.tpl.py")]
pub struct EnumTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub variants: Cow<'a, [EnumVariantTemplate<'a>]>,
}

#[derive(Clone)]
pub struct FieldTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: Cow<'a, str>,
}

#[derive(Clone)]
pub enum TypeTemplate<'a> {
    Unit,
    Tuple(Cow<'a, [Cow<'a, str>]>),
    Named {
        fields: Cow<'a, [FieldTemplate<'a>]>,
    },
}

#[derive(Template)]
#[template(escape = "none", path = "python/struct.tpl.py")]
pub struct StructTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
}
