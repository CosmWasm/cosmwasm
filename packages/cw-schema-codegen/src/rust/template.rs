use askama::Template;
use std::borrow::Cow;

#[derive(Clone)]
pub struct EnumVariantTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub serde_rename: Option<Cow<'a, str>>,
    pub ty: TypeTemplate<'a>,
}

#[derive(Template)]
#[template(escape = "none", path = "rust/enum.tpl.rs")]
pub struct EnumTemplate<'a> {
    pub add_allow: bool,
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub variants: Cow<'a, [EnumVariantTemplate<'a>]>,
}

#[derive(Clone)]
pub struct FieldTemplate<'a> {
    pub name: Cow<'a, str>,
    pub defaulting: bool,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: Cow<'a, str>,
}

#[derive(Clone)]
pub enum TypeTemplate<'a> {
    Unit,
    Tuple(Cow<'a, [Cow<'a, str>]>),
    Named {
        fields: Cow<'a, [FieldTemplate<'a>]>,
    },
}

#[derive(Template)]
#[template(escape = "none", path = "rust/struct.tpl.rs")]
pub struct StructTemplate<'a> {
    pub add_allow: bool,
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
}
