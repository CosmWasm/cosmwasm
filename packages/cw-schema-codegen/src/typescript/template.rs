use askama::Template;
use std::borrow::Cow;

#[derive(Clone)]
pub struct EnumVariantTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
}

#[derive(Template)]
#[template(escape = "none", path = "typescript/enum.tpl.ts")]
pub struct EnumTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub variants: Cow<'a, [EnumVariantTemplate<'a>]>,
    pub add_imports: bool,
}

#[derive(Clone)]
pub struct FieldTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: Cow<'a, str>,
}

#[derive(Clone)]
pub enum TypeTemplate<'a> {
    Unit,
    Tuple(Cow<'a, [Cow<'a, str>]>),
    Named {
        fields: Cow<'a, [FieldTemplate<'a>]>,
    },
}

#[derive(Template)]
#[template(escape = "none", path = "typescript/struct.tpl.ts")]
pub struct StructTemplate<'a> {
    pub name: Cow<'a, str>,
    pub docs: Cow<'a, [Cow<'a, str>]>,
    pub ty: TypeTemplate<'a>,
    pub add_imports: bool,
}
