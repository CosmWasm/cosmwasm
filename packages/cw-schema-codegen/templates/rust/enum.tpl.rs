// This code is @generated by cw-schema-codegen. Do not modify this manually.

{% if add_allow %}
#![allow(non_camel_case_types)]
{% endif %}

{% for doc in docs %}
    #[doc = "{{ doc }}"]
{% endfor %}

#[cosmwasm_schema::cw_serde]
pub enum {{ name }} {
{% for variant in variants %}
{% for doc in variant.docs %}
    #[doc = "{{ doc }}"]
{% endfor %}

{% match variant.serde_rename %}
{% when Some with (rename) %}
    #[serde(rename = "{{ rename }}")]
{% when None %}
{% endmatch %}

    {{ variant.name }}
{% match variant.ty %}
{% when TypeTemplate::Unit %}
{% when TypeTemplate::Tuple with (types) %}
        (
            {{ types|join(", ") }}
        )
{% when TypeTemplate::Named with { fields } %}
        {
        {% for field in fields %}
        {% for doc in field.docs %}
            #[doc = "{{ doc }}"]
        {% endfor %}
            {{ field.name }}: {{ field.ty }},
        {% endfor %}
        }
{% endmatch %}
        ,
{% endfor %}
}