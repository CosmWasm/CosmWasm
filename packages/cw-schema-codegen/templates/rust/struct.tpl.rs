// This code is @generated by cw-schema-codegen. Do not modify this manually.

#![allow(non_camel_case_types)]

{% for doc in docs %}
#[doc = "{{ doc }}"]
{% endfor %}

#[cosmwasm_schema::cw_serde]
pub struct {{ name }}

{% match ty %}
{% when TypeTemplate::Unit %}
;
{% when TypeTemplate::Tuple with (types) %}
(
    {{ types|join(", ") }}
);
{% when TypeTemplate::Named with { fields } %}
{
{% for field in fields %}
{% for doc in field.docs %}
    #[doc = "{{ doc }}"]
{% endfor %}

    {{ field.name }}: {{ field.ty }},
{% endfor %}
}
{% endmatch %}
