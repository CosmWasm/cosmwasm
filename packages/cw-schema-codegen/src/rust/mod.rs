use self::template::{
    EnumTemplate, EnumVariantTemplate, FieldTemplate, StructTemplate, TypeTemplate,
};
use heck::ToPascalCase;
use std::{borrow::Cow, io};

pub mod template;

fn expand_node_name<'a>(
    schema: &'a cw_schema::SchemaV1,
    node: &'a cw_schema::Node,
) -> Cow<'a, str> {
    match node.value {
        cw_schema::NodeType::Array { items } => {
            let items = &schema.definitions[items];
            format!("Vec<{}>", expand_node_name(schema, items)).into()
        }
        cw_schema::NodeType::Float => "f32".into(),
        cw_schema::NodeType::Double => "f64".into(),
        cw_schema::NodeType::Boolean => "bool".into(),
        cw_schema::NodeType::String => "String".into(),
        cw_schema::NodeType::Integer { signed, precision } => {
            let ty = if signed { "i" } else { "u" };
            format!("{ty}{precision}").into()
        }
        cw_schema::NodeType::Binary => "Vec<u8>".into(),

        cw_schema::NodeType::Boxed { inner } => {
            let inner = &schema.definitions[inner];
            format!("::alloc::boxed::Box<{}>", expand_node_name(schema, inner)).into()
        }
        cw_schema::NodeType::Optional { inner } => {
            let inner = &schema.definitions[inner];
            format!(
                "::core::option::Option<{}>",
                expand_node_name(schema, inner)
            )
            .into()
        }

        cw_schema::NodeType::Map {
            ref kind,
            key,
            value,
        } => {
            let key = expand_node_name(schema, &schema.definitions[key]);
            let value = expand_node_name(schema, &schema.definitions[value]);

            match kind {
                cw_schema::MapKind::BTree => {
                    format!("::alloc::collections::BTreeMap<{key}, {value}>")
                }
                cw_schema::MapKind::Hash => format!("::std::collections::HashMap<{key}, {value}>"),
                _ => unimplemented!(),
            }
            .into()
        }
        cw_schema::NodeType::Struct(..) => node.name.as_ref().into(),
        cw_schema::NodeType::Tuple { ref items } => {
            let items = items
                .iter()
                .map(|item| expand_node_name(schema, &schema.definitions[*item]))
                .collect::<Vec<_>>()
                .join(", ");

            format!("({})", items).into()
        }
        cw_schema::NodeType::Enum { .. } => node.name.as_ref().into(),

        cw_schema::NodeType::Decimal {
            precision: _,
            signed: _,
        } => {
            // ToDo: Actually use a decimal type here
            "String".into()
        }
        cw_schema::NodeType::Address => "cosmrs::AccountId".into(),
        cw_schema::NodeType::Checksum => "cosmrs::tendermint::Hash".into(),
        cw_schema::NodeType::HexBinary => {
            // ToDo: Actually use a hex-encoded binary type here
            "String".into()
        }
        cw_schema::NodeType::Timestamp => "cosmrs::tendermint::Time".into(),
        cw_schema::NodeType::Unit => "()".into(),
    }
}

fn prepare_docs(desc: Option<&str>) -> Cow<'_, [Cow<'_, str>]> {
    desc.map(|desc| {
        desc.lines()
            .map(|line| line.replace('"', "\\\"").into())
            .collect()
    })
    .unwrap_or(Cow::Borrowed(&[]))
}

pub fn process_node<O>(
    output: &mut O,
    schema: &cw_schema::SchemaV1,
    node: &cw_schema::Node,
    add_allow: bool,
) -> io::Result<()>
where
    O: io::Write,
{
    match node.value {
        cw_schema::NodeType::Struct(ref sty) => {
            let structt = StructTemplate {
                add_allow,
                name: node.name.clone(),
                docs: prepare_docs(node.description.as_deref()),
                ty: match sty {
                    cw_schema::StructType::Unit => TypeTemplate::Unit,
                    cw_schema::StructType::Named { ref properties } => TypeTemplate::Named {
                        fields: properties
                            .iter()
                            .map(|(name, prop)| FieldTemplate {
                                name: Cow::Borrowed(name),
                                defaulting: prop.defaulting,
                                docs: prepare_docs(prop.description.as_deref()),
                                ty: expand_node_name(schema, &schema.definitions[prop.value]),
                            })
                            .collect(),
                    },
                    cw_schema::StructType::Tuple { ref items } => TypeTemplate::Tuple(
                        items
                            .iter()
                            .map(|item| expand_node_name(schema, &schema.definitions[*item]))
                            .collect(),
                    ),
                },
            };

            writeln!(output, "{structt}")?;
        }
        cw_schema::NodeType::Enum { ref cases, .. } => {
            let enumm = EnumTemplate {
                add_allow,
                name: node.name.clone(),
                docs: prepare_docs(node.description.as_deref()),
                variants: cases
                    .iter()
                    .map(|(name, case)| EnumVariantTemplate {
                        name: name.to_pascal_case().into(),
                        docs: prepare_docs(case.description.as_deref()),
                        serde_rename: Some(name.clone()),
                        ty: match case.value {
                            cw_schema::EnumValue::Unit => TypeTemplate::Unit,
                            cw_schema::EnumValue::Tuple { ref items } => {
                                let items = items
                                    .iter()
                                    .map(|item| {
                                        expand_node_name(schema, &schema.definitions[*item])
                                    })
                                    .collect();

                                TypeTemplate::Tuple(items)
                            }
                            cw_schema::EnumValue::Named { ref properties, .. } => {
                                TypeTemplate::Named {
                                    fields: properties
                                        .iter()
                                        .map(|(name, prop)| FieldTemplate {
                                            name: Cow::Borrowed(name),
                                            defaulting: prop.defaulting,
                                            docs: prepare_docs(prop.description.as_deref()),
                                            ty: expand_node_name(
                                                schema,
                                                &schema.definitions[prop.value],
                                            ),
                                        })
                                        .collect(),
                                }
                            }
                        },
                    })
                    .collect(),
            };

            writeln!(output, "{enumm}")?;
        }
        _ => (),
    }

    Ok(())
}
