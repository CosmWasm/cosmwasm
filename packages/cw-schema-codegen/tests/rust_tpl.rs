use askama::Template;
use cw_schema_codegen::rust::template::{
    EnumTemplate, EnumVariantTemplate, FieldTemplate, StructTemplate, TypeTemplate,
};
use std::borrow::Cow;

#[test]
fn simple_enum() {
    let tpl = EnumTemplate {
        add_allow: true,
        name: Cow::Borrowed("Simple"),
        docs: Cow::Borrowed(&[Cow::Borrowed("Simple enum")]),
        variants: Cow::Borrowed(&[
            EnumVariantTemplate {
                name: Cow::Borrowed("One"),
                docs: Cow::Borrowed(&[Cow::Borrowed("One variant")]),
                serde_rename: None,
                ty: TypeTemplate::Unit,
            },
            EnumVariantTemplate {
                name: Cow::Borrowed("Two"),
                docs: Cow::Borrowed(&[Cow::Borrowed("Two variant")]),
                serde_rename: None,
                ty: TypeTemplate::Unit,
            },
        ]),
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn complex_enum() {
    let tpl = EnumTemplate {
        add_allow: true,
        name: Cow::Borrowed("Complex"),
        docs: Cow::Borrowed(&[Cow::Borrowed("Complex enum")]),
        variants: Cow::Borrowed(&[
            EnumVariantTemplate {
                name: Cow::Borrowed("One"),
                docs: Cow::Borrowed(&[Cow::Borrowed("One variant")]),
                serde_rename: None,
                ty: TypeTemplate::Tuple(Cow::Borrowed(&[Cow::Borrowed("u64")])),
            },
            EnumVariantTemplate {
                name: Cow::Borrowed("Two"),
                docs: Cow::Borrowed(&[Cow::Borrowed("Two variant")]),
                serde_rename: None,
                ty: TypeTemplate::Named {
                    fields: Cow::Borrowed(&[
                        FieldTemplate {
                            name: Cow::Borrowed("a"),
                            defaulting: false,
                            docs: Cow::Borrowed(&[Cow::Borrowed("Field a")]),
                            ty: Cow::Borrowed("u64"),
                        },
                        FieldTemplate {
                            name: Cow::Borrowed("b"),
                            defaulting: true,
                            docs: Cow::Borrowed(&[Cow::Borrowed("Field b")]),
                            ty: Cow::Borrowed("String"),
                        },
                    ]),
                },
            },
        ]),
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn empty_enum() {
    let tpl = EnumTemplate {
        add_allow: true,
        name: Cow::Borrowed("Empty"),
        docs: Cow::Borrowed(&[Cow::Borrowed("Empty enum")]),
        variants: Cow::Borrowed(&[]),
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn empty_struct() {
    let tpl = StructTemplate {
        add_allow: true,
        name: Cow::Borrowed("Empty"),
        docs: Cow::Borrowed(&[Cow::Borrowed("Empty struct")]),
        ty: TypeTemplate::Unit,
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn tuple_struct() {
    let tpl = StructTemplate {
        add_allow: true,
        name: Cow::Borrowed("Tuple"),
        docs: Cow::Borrowed(&[Cow::Borrowed("Tuple struct")]),
        ty: TypeTemplate::Tuple(Cow::Borrowed(&[
            Cow::Borrowed("u64"),
            Cow::Borrowed("String"),
        ])),
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn named_struct() {
    let tpl = StructTemplate {
        add_allow: true,
        name: Cow::Borrowed("Named"),
        docs: Cow::Borrowed(&[Cow::Borrowed("Named struct")]),
        ty: TypeTemplate::Named {
            fields: Cow::Borrowed(&[
                FieldTemplate {
                    name: Cow::Borrowed("a"),
                    defaulting: false,
                    docs: Cow::Borrowed(&[Cow::Borrowed("Field a")]),
                    ty: Cow::Borrowed("u64"),
                },
                FieldTemplate {
                    name: Cow::Borrowed("b"),
                    defaulting: true,
                    docs: Cow::Borrowed(&[Cow::Borrowed("Field b")]),
                    ty: Cow::Borrowed("String"),
                },
            ]),
        },
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}
