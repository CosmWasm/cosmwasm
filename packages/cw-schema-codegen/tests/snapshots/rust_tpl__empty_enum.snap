---
source: packages/cw-schema-codegen/tests/rust_tpl.rs
expression: rendered
---
// This code is @generated by cw-schema-codegen. Do not modify this manually.

#![allow(non_camel_case_types)]


    #[doc = "Empty enum"]


#[cosmwasm_schema::cw_serde]
pub enum Empty {

}
